### PR TITLE
Stunting  confusion matrix

### PR DESCRIPTION
**Pull request by prajurock**
# Description

A confusion matrix of stunting diagnosis.

User story # ([742](https://dev.azure.com/cgmorg/ChildGrowthMonitor/_workitems/edit/742))

# Result
stunting_diagnosis_q3-depthmap-plaincnn-height-95k_1610709869_2e00a6ef- Run 4
stunting_diagnosis_q3-depthmap-plaincnn-height-95k_1610709869_2e00a6ef

# Todo
#245 Make cgmzscore test work

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests
- [ ] New and existing unit tests pass locally with my changes
- [ ] My Code is review by 1 People

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -21,3 +21,6 @@ pyntcloud==0.1.3
 torch==1.7.0
 torchvision==0.8.1
 bunch==1.0.1
+seaborn
+scikit-learn
+cgmzscore
```
Comment (markus-hinsche) [inline requirements.txt:26]: Please pin the versions that you see in the logs when this gets installed

Comment (markus-hinsche) [inline requirements.txt:26]: There is a circular dependency here which is tricky: This repo builds cgmzscore. But now this repo also needs cgmzscore. 

Please try to remove this cgmzscore dependency

File: src/common/evaluation/QA/environment.yml [hunk 2]
```diff
@@ -17,4 +17,4 @@ dependencies:
     - glob2==0.7
     - tensorflow==2.3.1
     - tensorflow-addons==0.11.2
-    - bunch==1.0.1
+    - bunch==1.0.1
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/environment.yml:20]: Please undo this change

File: src/common/evaluation/QA/eval_depthmap_models/eval_main.py [hunk 3]
```diff
@@ -99,7 +99,10 @@
         "opencv-python==4.1.2.30",
         "matplotlib",
         "tensorflow-addons==0.11.2",
-        "bunch==1.0.1"
+        "bunch==1.0.1",
+        "seaborn",
+        "cgmzscore",
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/eval_main.py:104]: Here it is fine to have cgmzscore, however the versions should ideally be pinned

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py [hunk 4]
```diff
@@ -10,9 +11,12 @@
 import tensorflow as tf
 from azureml.core import Experiment, Run, Workspace
 from bunch import Bunch
+from sklearn.metrics import confusion_matrix
 import matplotlib
 matplotlib.use('Agg')
 import matplotlib.pyplot as plt  # noqa: E402
+from cgmzscore import Calculator  # noqa: E402
+import seaborn as sns  # noqa: E402
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/utils.py:19]: To clean up the code, put these dependencies before `matplotlib.use('Agg')` and you can remove `  # noqa: E402`

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py [hunk 5]
```diff
@@ -295,6 +299,35 @@ def draw_age_scatterplot(df_: pd.DataFrame, csv_out_fpath: str):
...
 
 
+def draw_stunting_diagnosis(df: pd.DataFrame, csv_out_fpath: str):
+    predicted_stunting = []
+    actual_stunting = []
+    not_processedData = []
+    for index, row in df.iterrows():
+        sex = 'M' if row[COLUMN_NAME_SEX] == SEX_DICT['male'] else 'F'
+
+        if row['GT'] > 45 and row['GT'] <= 120 and row['predicted'] > 45 and row['predicted'] <= 120 and row[COLUMN_NAME_AGE] <= 1856.0:
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/utils.py:309]: To make this line more readible:
```suggestion
        if 45< row['GT'] <= 120 and 45 < row['predicted'] <= 120 and row[COLUMN_NAME_AGE] <= 1856.0:
```

Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/utils.py:309]: There are magic numbers in this line: `45`, `120`, `1856.0`. 
Please define them as a variable and give them a suited name.

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py [hunk 6]
```diff
@@ -295,6 +299,35 @@ def draw_age_scatterplot(df_: pd.DataFrame, csv_out_fpath: str):
...
+            v = json.loads(Calculator().zScore_withclass(weight="0", muac="0",
+                                                         age_in_days=int(row[COLUMN_NAME_AGE]), sex=sex, height=row['GT']))
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/utils.py:311]: This code is a bit hard to read. Consider doing this in multiple steps:
```
aaa = age_in_days=int(row[COLUMN_NAME_AGE])
bbb = Calculator().zScore_withclass(weight="0", muac="0", aaa, sex=sex, height=row['GT'])
v = json.loads(bbb)
```

Please find good names for the variables `aaa`, `bbb`, and `v`

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py [hunk 7]
```diff
@@ -295,6 +299,35 @@ def draw_age_scatterplot(df_: pd.DataFrame, csv_out_fpath: str):
...
+            actual_stunting.append(v['Class_HFA'])
+            k = json.loads(Calculator().zScore_withclass(weight="0", muac="0",
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/utils.py:313]: same as above

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py [hunk 8]
```diff
@@ -295,6 +299,35 @@ def draw_age_scatterplot(df_: pd.DataFrame, csv_out_fpath: str):
...
+    data = confusion_matrix(actual_stunting, predicted_stunting)
+    df_cm = pd.DataFrame(data, columns=np.unique(actual_stunting), index=np.unique(actual_stunting))
+    df_cm.index.name = 'Actual'
+    df_cm.columns.name = 'Predicted'
+    plt.figure(figsize=(10, 7))
+    sns.set(font_scale=0.8)  # for label size
+    sns.heatmap(df_cm, cmap="Blues", annot=True, annot_kws={"size": 16}, fmt="d")  # font size
+    plt.title("Stunting Diagnosis")
+    Path(csv_out_fpath).parent.mkdir(parents=True, exist_ok=True)
+    plt.savefig(csv_out_fpath)
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/utils.py:327]: Should csv_out_fpath be png_out_fpath?

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py [hunk 9]
```diff
@@ -295,6 +299,35 @@ def draw_age_scatterplot(df_: pd.DataFrame, csv_out_fpath: str):
...
+            predicted_stunting.append(k['Class_HFA'])
+        else:
+            not_processedData.append(row['qrcode'])
+    data = confusion_matrix(actual_stunting, predicted_stunting)
+    df_cm = pd.DataFrame(data, columns=np.unique(actual_stunting), index=np.unique(actual_stunting))
+    df_cm.index.name = 'Actual'
+    df_cm.columns.name = 'Predicted'
+    plt.figure(figsize=(10, 7))
+    sns.set(font_scale=0.8)  # for label size
+    sns.heatmap(df_cm, cmap="Blues", annot=True, annot_kws={"size": 16}, fmt="d")  # font size
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/utils.py:324]: Do we really need to add seaborn? This introduces an extra new plotting library (not sure if this is reason enough to introduce a whole new library)? 
Could we do this only with matplotlib?

File: pytest.ini [hunk 10]
```diff
@@ -1,4 +1,4 @@
 [pytest]
-addopts = -p no:warnings --ignore=src/data_utils/dataset_generation/tests/test_create_dataset.py
+addopts = -p no:warnings --ignore=src/data_utils/dataset_generation/tests/test_create_dataset.py --ignore=src/common/zscore/cgmzscore/test_cgmzscore.py
```
Comment (markus-hinsche) [inline pytest.ini:2]: Why was this changed? These tests should run. I suggest to revert this change

**Review by markus-hinsche — COMMENTED** [review]
Thanks for the PR 💪. The confusion matrix looks reasonable. Good job. 
I made a few suggestion how to make the code more readable, most of them should be easy to include

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py (hunk 4 above)
Comment (prajurock) [reply src/common/evaluation/QA/eval_depthmap_models/src/utils.py:19]: cgmzscore also haveimport `matplotlib.pyplot as plt ` used. That's why doing this otherwise pytest fails

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py (hunk 8 above)
Comment (prajurock) [reply src/common/evaluation/QA/eval_depthmap_models/src/utils.py:327]: What you suggest?

File: pytest.ini (hunk 10 above)
Comment (prajurock) [reply pytest.ini:2]: Will resolve this in next pr

**Comment by prajurock** [comment]
https://dev.azure.com/cgmorg/ChildGrowthMonitor/_build/results?buildId=2367&view=results

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py (hunk 8 above)
Comment (markus-hinsche) [reply src/common/evaluation/QA/eval_depthmap_models/src/utils.py]: If you save a png file, then the name "csv_out_fpath" is misleading

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py (hunk 9 above)
Comment (prajurock) [reply src/common/evaluation/QA/eval_depthmap_models/src/utils.py]: removed

File: src/common/evaluation/QA/eval_depthmap_models/src/utils.py [hunk 11]
```diff
@@ -290,8 +299,41 @@ def draw_age_scatterplot(df_: pd.DataFrame, csv_out_fpath: str):
...
+def draw_stunting_diagnosis(df: pd.DataFrame, png_out_fpath: str):
+    """Draw stunting Confusion Matrix
+
+    Args:
+        df_: Dataframe with columns: qrcode, scantype, COLUMN_NAME_AGE, GT, predicted
+        png_out_fpath: File path where plot image will be saved
+    """
+    predicted_stunting = []
+    actual_stunting = []
+    not_processedData = []
```
Comment (markus-hinsche) [inline src/common/evaluation/QA/eval_depthmap_models/src/utils.py:316]:
```suggestion
    not_processed_data = []
```

**Review by markus-hinsche — APPROVED** [review]
LGTM 🔥 You covered all the comments very well. The code quality is quite good now!

**Comment by markus-hinsche** [comment]
The labels in the bottom are hard to read:

image

**Comment by markus-hinsche** [comment]
The labels on the left are cut off

image

**Comment by markus-hinsche** [comment]
One idea could be to also have percentages below each count. But we can do this later
